javascript/react-js/state-and-props.md

```diff
@@ -0,0 +1,240 @@
+
+As you can see when you are passing many properties or functions to a component it can get quite exhausting to always refere to them with `this.props.somePropery`. Destructuring for the rescue. We can aternatively write the above as follows:
+
+```javascript
+// MyComponent.js
+
+import React, { Component } from "react";
+
+class MyComponent extends Component {
+  constructor(props) {
+    super(props);
+  }
+
+  render() {
+    const { title, onButtonClicked } = this.props;
+
+    return (
+      <div>
+        <h1>{title}</h1>
+        <button onClick={onButtonClicked}>Click Me!</button>
+      </div>
+    );
+  }
+}
+
+export default MyComponent;
+```
+
+Here we are destructuring `title` and `onButtonClicked` from `this.props`, which lets us refer to them with just their names. Make sure to destructure the render method when using class components. In functional components you would destructure outside of the return statement.
+
+Great! So far so good, you hopefully have understood the purpose and usage of `props` in React. Isn't that difficult is it?!
+
+### State
+
+Another very important concept in React is state. In short, all values that can change over time are handled through state.
+
+And that's the way to define `state` in React:
+
+```javascript
+import React, { Components } from "react";
+
+class App extends Component {
+  constructor() {
+    super();
+
+    this.state = {
+      count: 0,
+    };
+  }
+
+  countUp() {
+    this.setState({
+      count: this.state.count + 1,
+    });
+  }
+
+  render() {
+    return (
+      <div>
+        <button onClick={this.countUp}>Click Me!</button>
+        <p>{this.state.count}</p>
+      </div>
+    );
+  }
+}
+```
+
+In the above component we declared a state. You **always** declare state in the constructor in a class component. Remember that for functional components it works a little bit different. The `setState` method we call sets the state to a new value. In React you should **never** change state directly. What does that mean? Never do something like this: `this.state.count = 3`, instead always use the `setState` method React provides you to modify the state. Really keep that in mind, it can save you a lot of debugging when you are getting started with React.
```

Comment:
I believe it's worth directly mentioning "immutability" in regard to never changing state directly and how it can lead to unexpected behavior.

Reply:
I totally agree. I already included it.